### ISSUE TITLE
Building and Deploying a Docker Image for a React-Django Web App on Ubuntu


### ISSUE THREAD

**Issue by rupeshakhade** [rupeshakhade/Docker_Project#1]
# Step 1: Connect to an EC2 instance.
 
Connect to your ec2 instance using it's pem file & public DNS

---->>ssh -i "< pem file > ubuntu@ < public DNS >

![Screenshot (541).jpg](https://user-images.githubusercontent.com/20862722/233465556-ef4ad6d7-367e-4ee9-bc03-4939d5b19c09.jpg)

# Step 2: Update the package.

Update the package manager by using below command

----->> sudo apt-get update

# Step 3: install docker 

Install docker using the below command 

----->> sudo apt-get install docker.io

Check the docker version and status.

We can check the docker version and status to ensure that it was properly installed or not.

To verify the Docker Version and Docker status use below commands

----->> systemctl status docker

----->> docker --version.

# **Step** 4: Create directory.

We will create one directory and clone the code from github

----->> mkdir projects

Enter the projects folder

----->> cd projects/

# Step 5: Clone the repository from Github.

In this step, we clone the code for the Django-React demo app from Github using the “git clone” command. The code will be downloaded and stored in a directory.

-----> git clone https://github.com/rupeshakhade/react_django_demo_app.git

![Screenshot (542).jpg](https://user-images.githubusercontent.com/20862722/233468125-95e45a98-b7ce-417e-bc94-99bc60486249.jpg)

# **Step 6: Check the files and understand the requirements**.

After cloning the code, we can check the files in the directory and understand the requirements needed to run the code. The “ls” command can be used to check the files, and we can refer to the README file and manage.py file to understand the code and its dependencies.

 · After doing the clone, a directory will be created with a name called react_django_demo_app

· Go to that directory and do ls  to check files.

Commands: 
----->>cd react_django_demo1_app
----->>ls

![Screenshot (543).jpg](https://user-images.githubusercontent.com/20862722/233472507-c969ad12-a5c9-4619-93ef-1a6dd939abe9.jpg)

README file — This is the where developer gives steps to run this code.

Here developer has not given much information about the code so we will other files to understand the code.

Manage.py file — Generally this is the file that contains the main things and we have to run this file by seeing the extension (.py) we can say this is Python project and we need to install Python to run is code.

Requirements.txt file — This is file we will get to know all requirements we need to run this code. Here in this project, we need Django also to run it.

# Step 7: Create a Dockerfile.

A Dockerfile is a text file that includes the necessary commands for building a Docker image. In this step, we create a Dockerfile using the “vim Dockerfile” command,

----->>vim Dockerfile

and  include the necessary commands such as “FROM”, “RUN”, “COPY”, “EXPOSE”, and “CMD”. These commands are used to install packages, create directories, set environment variables, and specify the command to be run when the container is created.

Docker file — It is a text file that has all commands which need to be run for building a given image.

![Screenshot (545).jpg](https://user-images.githubusercontent.com/20862722/233474867-843b6409-5507-415f-8e04-a2b9a78c56b5.jpg)

FROM: This command specifies the base image that the new image will be built on top of. The base image can be an official image from the Docker Hub or a custom image from a private repository.

COPY: This command copies files or directories from the host machine to the container.

EXPOSE: This command informs Docker that the container listens on the specified network ports at runtime.

CMD: This command specifies the command that will be run when a container is created from the image.

# **Step 8: Build a docker image**.

After creating the Dockerfile, we can build the Docker image using the “docker build” command. This command builds an image from the Dockerfile and tags it with a name and version.

Docker image — They are executable packages, bundled with application code & dependencies, software packages, etc. for the purpose of creating containers. Docker images can be deployed to any docker environment and the container can be spun up there to run the application.

----->>docker build . -t react-django-app:latest

![Screenshot (547).jpg](https://user-images.githubusercontent.com/20862722/233476389-bfce0bc2-4e25-4669-a7ca-487b0b961352.jpg)

After building the Docker image, we can check the images using the “docker images” command. This command lists all the images that have been built on the Docker environment.

![Screenshot (548).jpg](https://user-images.githubusercontent.com/20862722/233476716-aa08877d-9695-410f-bbbf-c535f132597f.jpg)

# **Step 9: Create a Docker container**.

Once we have built the Docker image, we can create a Docker container using the “docker run” command. This command creates a container from the image and maps the container’s port to the host’s port.

Docker container — Container hold the entire package that is needed to run a application, libraries and system dependencies.

- Now using above images we will create a container by using command

----->>docker run -d -p 8001:8001 react-django-app:latest

:   -d is daemon mode or detach mode

:   -p is port mapping of container port 8000 to  instance port 8000

![Screenshot (550).jpg](https://user-images.githubusercontent.com/20862722/233480546-685c5c97-0ace-4288-aa15-31ace6f07b22.jpg)

- We can check the list of Docker containers using the “docker ps” command. This command lists all the running Docker containers.

![Screenshot (551).jpg](https://user-images.githubusercontent.com/20862722/233481253-6d6b6432-0501-4289-9833-3383b784b96f.jpg) 

# **Step 10: Port opening**

Open port 8001 in the instance's security inbound rule Check if the application is running.

Before checking the application on a web browser, we need to open port 8001 in the instance’s security inbound rule to allow traffic to flow to the container.

![Screenshot (555).jpg](https://user-images.githubusercontent.com/20862722/233482290-3711f5f2-7abc-43cc-80d7-e5b6f65b572c.jpg)

# **Step 11: Check if the application is running**

Finally, we can copy the public IP of the instance to the web browser and check if the application is running. If everything has been set up correctly, the Django-React demo app should be up and running.

![Screenshot (552).png](https://user-images.githubusercontent.com/20862722/233482720-6f2bfa37-17ce-40f8-bca5-71f471c69990.png)

In summary, setting up a Docker container for a Django-React demo app on Ubuntu involves updating the packet manager, installing Docker, cloning the code from Github, understanding the code and its requirements, creating a Dockerfile, building the Docker image, creating a Docker container, checking the list of Docker containers, opening the required port in the instance’s security inbound rule, and checking if the application is running on a browse.

![500_F_283424830_9sf5qvuEMPw1fvPIM0g3vmKJHpfsgJwf.jpg](https://user-images.githubusercontent.com/20862722/233486907-81338ea0-39cf-4a07-b00b-954e053515cf.jpg)

Happy learnings😃🤓

**Comment by prathams99** [comment]
Zaaatu zavnya he je tu issues madhe lihilas te README madhe pahije
